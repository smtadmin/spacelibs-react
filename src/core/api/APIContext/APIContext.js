/*
 * File: /src/core/api/APIContext/APIContext.js
 * Version: 0.0.2
 * Project: @siliconmtn/spacelibs-react
 * Description: API Context
 * File Created: Wednesday, 17th February 2021 4:49 pm
 * Author: tyler Gaffaney (tyler.gaffaney@siliconmtn.com)
 * -----
 * Last Modified: Thursday, 18th February 2021 3:36 pm
 * Modified By: tyler Gaffaney (tyler.gaffaney@siliconmtn.com>)
 * -----
 * Copyright 2021, Silicon Mountain Technologies, Inc.
 */

import React from "react";

const APIContext = React.createContext();
APIContext.displayName = "APIContext";

export default APIContext;
