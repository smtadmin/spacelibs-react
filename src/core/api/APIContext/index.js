/*
 * File: /src/core/api/APIContext/index.js
 * Version: 0.0.2
 * Project: @siliconmtn/spacelibs-react
 * Description: Index files are used to make exporting easier
 * File Created: Wednesday, 17th February 2021 4:49 pm
 * Author: tyler Gaffaney (tyler.gaffaney@siliconmtn.com)
 * -----
 * Last Modified: Wednesday, 3rd March 2021 8:46 am
 * Modified By: tyler Gaffaney (tyler.gaffaney@siliconmtn.com>)
 * -----
 * Copyright 2021, Silicon Mountain Technologies, Inc.
 */

import APIContext from './APIContext';

export default APIContext;