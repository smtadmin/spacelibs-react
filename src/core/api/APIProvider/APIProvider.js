/*
 * File: /src/core/api/P1APIProvider/P1APIProvider.js
 * Version: 0.0.2
 * Project: @siliconmtn/spacelibs-react
 * Description: This is the API Context provider
 * File Created: Wednesday, 17th February 2021 4:53 pm
 * Author: tyler Gaffaney (tyler.gaffaney@siliconmtn.com)
 * -----
 * Last Modified: Monday, 1st March 2021 1:10 pm
 * Modified By: tyler Gaffaney (tyler.gaffaney@siliconmtn.com>)
 * -----
 * Copyright 2021, Silicon Mountain Technologies, Inc.
 */

import React from "react";
import PropTypes from "prop-types";
import APIContext from "../APIContext";
import SecurityContext from "../../security/P1SecurityContext";
import BaseHTTPService from "@siliconmtn/spacelibs-js/core/io/BaseHTTPService";

/**
 * API Context provider
 * @param {object} props - Component props
 * @class APIProvider
 */
class APIProvider extends React.Component {

    /**
     * Constructs an APIProvider
     * @param {*} props - Component Props
     */
    constructor(props) {
        super(props);

        let config = {};

        if (this.props.baseUrl) {
            config.host = this.props.baseUrl;
		}

        this.state = {
            service: new BaseHTTPService(config),
		};
		
		this.read.bind(this);
		this.update.bind(this);
		this.insert.bind(this);
		this.delete.bind(this);
    }

	/**
	 * This adds authentication headers if user is authenticated
	 *
	 * @param {*} originalOptions - options passed in request
	 * @returns {*} newOptions with header
	 * @memberof APIProvider
	 */
	getOptionsWithAuth(originalOptions = {}) {
        const securityContext = this.context;
        if ( securityContext && securityContext.tokens && securityContext.tokens.accessToken ) {
            if (!originalOptions.headers) {
                originalOptions.headers = {};
            }

            originalOptions.headers = {
                Authorization: "Bearer " + securityContext.tokens.accessToken,
            };
		}
		return originalOptions;
    }

    /**
     * Request call for reading/selecting data.  Uses the `GET` HTTP Method.
     * @param {string} url - Location for the request to go to.  Can be absolute and relative (if you passed in baseUrl to the constructor)
     * @param {object} data - Data for the request
     * @param {Function} onSuccess - Handler for successful request.  Handler is passed response object.
     * @param {Function} onFailure - Handler for errors/failures with the request.  Handler is passed an error object.
     * @param {object} options - Additional options for this request.
     * @memberof BaseHTTPService
     */
    read(url, data, onSuccess, onFailure, options = {}) {
        this.state.service.read(
            url,
            data,
            onSuccess,
            onFailure,
            this.getOptionsWithAuth(options)
        );
    }

    /**
     * Request call for inserting data.  Uses the `POST` HTTP Method.
     * @param {string} url - Location for the request to go to.  Can be absolute and relative (if you passed in baseUrl to the constructor)
     * @param {object} data - Data for the request
     * @param {Function} onSuccess - Handler for successful request.  Handler is passed response object.
     * @param {Function} onFailure - Handler for errors/failures with the request.  Handler is passed an error object.
     * @param {object} options - Additional options for this request.
     * @memberof BaseHTTPService
     */
    insert(url, data, onSuccess, onFailure, options = {}) {
        this.state.service.insert(
            url,
            data,
            onSuccess,
            onFailure,
            this.getOptionsWithAuth(options)
        );
    }

    /**
     * Request call for deleting data.  Uses the `DELETE` HTTP Method.
     * @param {string} url - Location for the request to go to.  Can be absolute and relative (if you passed in baseUrl to the constructor)
     * @param {object} data - Data for the request
     * @param {Function} onSuccess - Handler for successful request.  Handler is passed response object.
     * @param {Function} onFailure - Handler for errors/failures with the request.  Handler is passed an error object.
     * @param {object} options - Additional options for this request.
     * @memberof BaseHTTPService
     */
    delete(url, data, onSuccess, onFailure, options = {}) {
        this.state.service.delete(
            url,
            data,
            onSuccess,
            onFailure,
            this.getOptionsWithAuth(options)
        );
    }

    /**
     * Request call for updating existing data.  Uses the `PATCH` HTTP Method.
     * @param {string} url - Location for the request to go to.  Can be absolute and relative (if you passed in baseUrl to the constructor)
     * @param {object} data - Data for the request
     * @param {Function} onSuccess - Handler for successful request.  Handler is passed response object.
     * @param {Function} onFailure - Handler for errors/failures with the request.  Handler is passed an error object.
     * @param {object} options - Additional options for this request.
     * @memberof BaseHTTPService
     */
    update(url, data, onSuccess, onFailure, options = {}) {
        this.state.service.update(
            url,
            data,
            onSuccess,
            onFailure,
            this.getOptionsWithAuth(options)
        );
    }

    /**
     * Renders components
     *
     * @returns {*} APIProvider component
     * @memberof APIProvider
     */
    render() {
        return (
            <APIContext.Provider
                value={{
                    read: this.read.bind(this),
                    update: this.update.bind(this),
                    insert: this.insert.bind(this),
                    delete: this.delete.bind(this),
                }}>
                {this.props.children}
            </APIContext.Provider>
        );
    }
}

/* Assinging context */
APIProvider.contextType = SecurityContext;

/**
 * Accepted props
 */
APIProvider.propTypes = {
    children: PropTypes.any,
    baseUrl: PropTypes.string,
};

export default APIProvider;
