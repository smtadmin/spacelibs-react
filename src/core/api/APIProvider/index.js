/*
 * File: /src/core/api/APIProvider/index.js
 * Version: 0.0.2
 * Project: @siliconmtn/spacelibs-react
 * Description: INSERT DESCRIPTION
 * File Created: Wednesday, 17th February 2021 4:49 pm
 * Author: tyler Gaffaney (tyler.gaffaney@siliconmtn.com)
 * -----
 * Last Modified: Wednesday, 17th February 2021 4:59 pm
 * Modified By: tyler Gaffaney (tyler.gaffaney@siliconmtn.com>)
 * -----
 * Copyright 2021, Silicon Mountain Technologies, Inc.
 */

import APIProvider from './APIProvider';

export default APIProvider;