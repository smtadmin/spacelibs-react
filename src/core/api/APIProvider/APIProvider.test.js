/*
 * File: /src/core/api/APIProvider/APIProvider.test.js
 * Version: 0.0.2
 * Project: @siliconmtn/spacelibs-react
 * Description: INSERT DESCRIPTION
 * File Created: Monday, 1st March 2021 8:47 am
 * Author: tyler Gaffaney (tyler.gaffaney@siliconmtn.com)
 * -----
 * Last Modified: Monday, 1st March 2021 2:03 pm
 * Modified By: tyler Gaffaney (tyler.gaffaney@siliconmtn.com>)
 * -----
 * Copyright 2021, Silicon Mountain Technologies, Inc.
 */

import APIProvider from "./APIProvider";
import APIContext from '../APIContext';
import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";

/**
 * Checks that the button renders and the current snapshot matches the previous one.
 */
it("Renders with no props", () => {
    const { baseElement } = render(
        <APIProvider />
    );
    expect(baseElement).toBeTruthy();
    expect(baseElement).toMatchSnapshot();
});

it("Renders with props", () => {
    const { baseElement } = render(<APIProvider baseUrl={"http://www.siliconmtn.com"} />);
    expect(baseElement).toBeTruthy();
    expect(baseElement).toMatchSnapshot();
});

it("Passes valid api information", () => {

	function Comp(props){
		expect(props.read).toBeDefined();
		expect(props.update).toBeDefined();
		expect(props.insert).toBeDefined();
		expect(props.delete).toBeDefined();
		
		return <div></div>;
	}

	const { baseElement } = render(<APIProvider>
		<APIContext.Consumer >
			{(value) => 
				<Comp {...value}/>
			}
		</APIContext.Consumer>
	</APIProvider>);

	expect(baseElement).toBeTruthy();
});