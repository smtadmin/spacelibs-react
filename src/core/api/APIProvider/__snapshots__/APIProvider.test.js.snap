// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders with no props 1`] = `
<body>
  <div />
</body>
`;

exports[`Renders with props 1`] = `
<body>
  <div />
</body>
`;
